lib/ansible/plugins/inventory/tower.py

```diff
@@ -0,0 +1,201 @@
+# Copyright (c) 2018 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: tower
+    plugin_type: inventory
+    authors:
+      - Matthew Jones (@matburt)
+      - Yunfan Zhang (@YunfanZhang42)
+    short_description: Ansible dynamic inventory plugin for Ansible Tower.
+    description:
+        - Imports inventories from Ansible Tower.
+        - Supports reading configuration from both YAML config file and environment variables.
+        - If reading from the YAML file, the file name must end with tower_inventory.(yml|yaml),
+          the path in the command would be /path/to/tower_inventory.(yml|yaml). If some arguments in the config file
+          are missing, this plugin will try to fill in missing arguments by reading from environment variables.
+        - If reading configurations from environment variables, the path in the command would be tower_inventory.
+    options:
+        tower_host:
+            description: The network address of your Ansible Tower host.
+            type: string
+            env:
+                - name: TOWER_HOST
+            required: True
+        tower_username:
+            description: The user that you plan to use to access inventories on tower.
+            type: string
+            env:
+                - name: TOWER_USERNAME
+            required: True
+        tower_password:
+            description: The password for tower_username.
+            type: string
+            env:
+                - name: TOWER_PASSWORD
+            required: True
```

Comment:
It also might be good to compare these options to the options in

https://github.com/ansible/ansible/blob/cdd21e2170b9a47d84624c8c2f792a66f2b0fe57/lib/ansible/utils/module_docs_fragments/tower.py#L23-L42

Reply:
@AlanCoding Thanks for the dropping in! It seems we are using same names for items in yaml config file, while keeping the environment variable names the same as the they were in inventory script, so we should be fine. 

I think we let the users to specify `tower_config_file` at run time, so this option is no longer needed right?

Reply:
> I think we let the users to specify tower_config_file at run time, so this option is no longer needed right?

What do you mean by this?

I think the current state breaks congruence with the tower modules in that it won't accept the `.tower_cli.cfg` file. I'm not going to push for that right now, because I don't have some kind of grand unification plan that would allow both to take the same forms of credentials, although you could _maybe_ just import the options processing code from the tower modules and have it work.

That may be too ambitious, and I'm braced for this being somewhat inconsistent of a specification.

If you look at my openstack usage example:

https://github.com/AlanCoding/Ansible-inventory-file-examples/blob/master/plugins/example_openstack/openstack.yml

You can see that the _native configuration file format_ is `clouds.yml`, and the _inventory file_ can do nothing more than point to the native configuration file.

Granted, I _hate_ this, but it seems to be the canonical way of writing inventory plugins. If you extrapolated this pattern, then yes, I think you would allow the entry of `tower_config_file` inside of the `tower_inventory.yml` file. @bcoca please correct me if I have horribly misunderstood your standards here.

Reply:
its fine to have a 'separate pre-existing file', I would just use it as a fallback if other methods are not populated and i would not create one if it were not already standard.

Reply:
i actually don't think you should force having the `tower_config_file` entry, but it could be posed as an option and default to the 'normal location'

Reply:
the openstack plugin is not really the best example of standards to follow, you might be better off looking at aws_ec2 which also has 'pre existing configs' via boto

Reply:
yes, I agree that _if_ `tower_config_file` is added as an option it should be optional. As a first-pass, maybe we don't need it, but have a plan for it in the future.

I still need to get synced up about the 2 methods of specification that this PR current does.

Reply:
So let me propose this, if just using the 'name' (no inventory source file) or the file contains a tower_config_file entry, use that first if they exist (in parse method):

```python
super(InventoryModule, self).parse(inventory, loader, path)
if self.read_all_configs_from_env_var: # I would change name to 'no config file supplied' or similar
   if os.path.exists('~/.tower_cli.cfg'):
      # pick up options from the file
      self._set_options_from_tower_cli('~/.tower_cli.ccfg')
    self.set_options() # this will update with env vars to override file or pick up others       
else:
   self._read_config_data(path)
   if self.get_option('tower_cli_config'): # check for user defined
        self._set_options_from_tower_cli(self.get_option('tower_cli_config')
 ...
```

Reply:
Our current decision is to keep this plugin simple and do not let it read from `tower_cli_config`. In the future we might add this option.